brian2genn/device.py

```diff
@@ -897,17 +950,74 @@ def generate_code_objects(self, writer):
                     self.header_files.append(
                         'code_objects/' + codeobj.name + '.h')
 
+    def generate_max_row_length_code_objects(self, writer):
+        # Generate data for non-constant values
+        code_object_defdefs = defaultdict(list)
+        code_object_defs = defaultdict(list)
+        for codeobj in itervalues(self.max_row_length_code_objects):
+            deflines= []
+            lines = []
+            for k, v in iteritems(codeobj.variables):
+                if isinstance(v, ArrayVariable):
+                    try:
+                        if isinstance(v, DynamicArrayVariable):
+                            if get_var_ndim(v) == 1:
+                                dyn_array_name = self.dynamic_arrays[v]
+                                array_name = self.arrays[v]
+                                # define the dynamic array
+                                line = 'vector<{c_type}> {dyn_array_name};'
+                                line = line.format(c_type=c_data_type(v.dtype),
+                                                   array_name=array_name,
+                                                   dyn_array_name=dyn_array_name)
+                                deflines.append(line)
+                                # do the const stuff
+                                line = '{c_type}* const {array_name} = &{dyn_array_name}[0];'
+                                line = line.format(c_type=c_data_type(v.dtype),
+                                                   array_name=array_name,
+                                                   dyn_array_name=dyn_array_name)
+                                lines.append(line)
+                                line = 'const int _num{k} = {dyn_array_name}.size();'
+                                line = line.format(k=k,
+                                                   dyn_array_name=dyn_array_name)
+                                lines.append(line)
+                        else:
+                            array_name = self.arrays[v]
+                            line = '{c_type} {array_name}[{size}];'
+                            line = line.format(c_type=c_data_type(v.dtype),
+                                               array_name=array_name,
+                                               size=v.size)
+                            deflines.append(line)
+                            lines.append('const int _num%s = %s;' % (k, v.size))
+                    except TypeError:
+                        pass
+            for line in lines:
+                # Sometimes an array is referred to by to different keys in our
+                # dictionary -- make sure to never add a line twice
+                if not line in code_object_defs[codeobj.name]:
```

Comment:
you could make ``code_object_defs`` and ``code_object_defdefs`` default dicts of sets rather than lists to have this behaviour implicitly

Reply:
Done that ... and it seems to work (obviously). But do check ...